File: pca/elixir/pca.exs
defmodule PCA do
  def pca(data) do
    # Standardize the data.
    mean = data.mean(axis=1)
    std = data.std(axis=1)
    data = data - mean
    data = data / std

    # Calculate the covariance matrix.
    covariance_matrix = data.dot(data, transpose=true) / (data.n - 1)

    # Calculate the eigenvalues and eigenvectors of the covariance matrix.
    eigenvalues, eigenvectors = covariance_matrix.eig()

    # Sort the eigenvalues and eigenvectors in descending order.
    eigenvalues, eigenvectors = eigenvalues |> Enum.sort_by(fn x -> -x end), eigenvectors |> Enum.sort_by(fn x -> -x end)

    # Return the principal components.
    eigenvectors
  end
end

File: pca/matlab/pca.m
clear all;
clc;

% Load the data.
data = [1, 2, 3, 4, 5];

% Standardize the data.
mean = sum(data) / length(data);
std = sqrt(sum((data - mean) .^ 2) / length(data));
for i = 1:length(data)
    data(i) = (data(i) - mean) / std;
end

% Calculate the covariance matrix.
cov = (data' * data) / (length(data) - 1);

% Calculate the eigenvalues and eigenvectors of the covariance matrix.
[eigs, vecs] = eig(cov);

% Sort the eigenvalues and eigenvectors in descending order.
[eigs, vecs] = sort(eigs, 'descend');

% Return the principal components.
principalComponents = vecs(:, 1:size(eigs, 1));

% Print the principal components.
disp(principalComponents);
